library/src/cache/patch_manager.rs

```diff
@@ -179,9 +190,52 @@ impl PatchManager {
             );
         }
 
+        // Verify that we haven't tried and failed to boot this patch.
+        if self.is_patch_last_attempted_patch(patch.number) {
+            match self.last_successful_boot_patch_number() {
+                Some(last_successful_patch_number)
+                    if last_successful_patch_number != patch.number =>
+                {
+                    // We've tried to boot this patch before and weren't successful,
+                    // so don't try again.
+                    bail!(
+                        "Already attempted and failed to boot patch {}",
+                        patch.number
+                    )
+                }
+                None => {
+                    // We've tried to boot from this patch before and didn't
+                    // succeed. Don't try again.
+                    bail!(
+                        "Already attempted and failed to boot patch {}",
+                        patch.number
+                    )
+                }
+                Some(_) => {
+                    // This is a known-good patch.
+                }
+            }
+        }
+
         Ok(())
     }
 
+    /// Whether the given patch number is the last one we attempted to boot
+    /// (whether it was successful or not).
+    fn is_patch_last_attempted_patch(&self, patch_number: usize) -> bool {
+        self.patches_state
+            .last_attempted_patch
```

Comment:
Would this also be self.patches_state.last_attempted_patch == Some(patch_number)?

Reply:
Oh, I see, last_attempted_patch isn't itself a number, it's a number + size.